crates/ty_python_semantic/resources/mdtest/subscript/tuple.md

```diff
@@ -69,8 +69,60 @@ def _(m: int, n: int):
     t[::0]  # error: [zero-stepsize-in-slice]
 
     tuple_slice = t[m:n]
-    # TODO: Should be `tuple[Literal[1, 'a', b"b"] | None, ...]`
-    reveal_type(tuple_slice)  # revealed: tuple[Unknown, ...]
+    reveal_type(tuple_slice)  # revealed: tuple[Literal[1, "a", b"b"] | None, ...]
+```
+
+## Slices of homogeneous and mixed tuples
+
+```toml
+[environment]
+python-version = "3.11"
+```
+
+```py
+from typing import Literal
+
+def homogeneous(t: tuple[str, ...]) -> None:
+    reveal_type(t[0])  # revealed: str
+    reveal_type(t[1])  # revealed: str
+    reveal_type(t[2])  # revealed: str
+    reveal_type(t[3])  # revealed: str
+
+    reveal_type(t[-1])  # revealed: str
+    reveal_type(t[-2])  # revealed: str
+    reveal_type(t[-3])  # revealed: str
+    reveal_type(t[-4])  # revealed: str
+
+def mixed(t: tuple[Literal[1], Literal[2], *tuple[str, ...], Literal[9], Literal[10]]) -> None:
+    reveal_type(t[0])  # revealed: Literal[1]
+    reveal_type(t[1])  # revealed: Literal[2]
+    reveal_type(t[2])  # revealed: str
+    reveal_type(t[3])  # revealed: str
+
+    reveal_type(t[-1])  # revealed: Literal[10]
+    reveal_type(t[-2])  # revealed: Literal[9]
+    reveal_type(t[-3])  # revealed: str
+    reveal_type(t[-4])  # revealed: str
```

Comment:
There were a lot of ecosystem failures of the form

```
Index 8 is out of bounds for tuple `tuple[Any, ...]` with length at least 0
```

I've fixed this in 2e41487de38c874efe06e0cf3e798f8295adebb5: indexing into a homogeneous or mixed tuple returns the variable-length portion if you index past any fixed-length prefix or suffix.

But I'm not convinced that's correct, since the spec says that a variable-length tuple is the _union_ of all possible lengths, not the gradual choice of them. That would suggest that we should treat this as a potential out-of-bounds error, since the index isn't valid for all possible elements (i.e. lengths) of the "union".

Reply:
Updated to include elements from the prefix/suffix as appropriate.